pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java

```diff
@@ -163,48 +190,64 @@ public synchronized void refreshSegment(String segment) {
       if (filterExpression == null) {
         return segments;
       }
-      Set<String> selectedSegments = new HashSet<>();
-      for (String segment : segments) {
-        PartitionInfo partitionInfo = _partitionInfoMap.get(segment);
-        if (partitionInfo == null || partitionInfo == INVALID_PARTITION_INFO || isPartitionMatch(filterExpression,
-            partitionInfo)) {
-          selectedSegments.add(segment);
-        }
-      }
-      return selectedSegments;
+      return pruneSegments(segments, (partitionInfo, cachedPartitionFunction) -> isPartitionMatch(filterExpression,
+        partitionInfo, cachedPartitionFunction));
     } else {
       // PQL
       FilterQueryTree filterQueryTree = RequestUtils.generateFilterQueryTree(brokerRequest);
       if (filterQueryTree == null) {
         return segments;
       }
-      Set<String> selectedSegments = new HashSet<>();
-      for (String segment : segments) {
+      return pruneSegments(segments, (partitionInfo, cachedPartitionFunction) -> isPartitionMatch(filterQueryTree, partitionInfo, cachedPartitionFunction));
+    }
+  }
+
+  private SelectedSegments pruneSegments(SelectedSegments segments, java.util.function.BiFunction<PartitionInfo, CachedPartitionFunction,  Boolean> partitionMatchLambda) {
+    // Some segments may not be refreshed/notified via Zookeeper yet, but broker may have found it
+    // in this case we need to include the new segments.
+    Set<String> selectedSegments = new HashSet<>();
+    CachedPartitionFunction cachedPartitionFunction = new CachedPartitionFunction();
+    if (segments.hasHash() && segments.getSegmentHash().equals(_allSegments.getSegmentHash())) {
```

Comment:
`TimeSegmentPruner` is always in front of `PartitionSegmentPruner` if configured, so this if statement won't be `true` if any segment is pruned.
It also adds a lot of complexity to the code as we always need to choose whether to calculate the hash.